art-decisions/proposals/2022-11-20-arcaflow-signals.md

```diff
@@ -0,0 +1,209 @@
+team. They have introduced us to their extremely powerful tool for executing workflows called
+[qDup](https://github.com/Hyperfoil/qDup). This tool not only allows people to write workflows, but also dynamically
+react to outputs of programs and publish signals on an internal messaging bus. Workflow parts can communicate with each
+other even while running, sending and waiting for signals.
+
+Arguably, qDup can do more than Arcaflow at this point. However, it requires a great deal of expertise and forethought
+as it doesn't have the safety guarantees that Arcaflow does, such as schemas or possible static analysis of the code.
+
+It is the aim of this proposal that Arcaflow should receive the signaling mechanism, allowing plugins to exchange
+messages during execution. This proposal incorporates the plan to create data streaming, as signals will allow plugins
+to transmit chunks of data.
+
+## Proposal
+
+Currently, Arcaflow allows the plugin to output only one of many possible outputs. This limits their usefulness,
+as the plugin must finish execution to transmit information.
+
+In this proposal, we transform the execution of Arcaflow by adding the ability to send and receive signals via signal
+channels. Each signal channel will have a schema and is declared by a plugin. Workflow authors can take these signals 
+and pipe them into other plugins that have declared they can receive signals.
+
+In a workflow, this may look as follows:
+
+```yaml
+steps:
+  foo:
+    ...
+  bar:
+    listen:
+      signal_listener_key_declared_by_plugin: !expr $.steps.foo.signals.something
+```
+
+In this case the `bar` plugin will have declared a signal listener called `signal_listener_key_declared_by_plugin`
+together with a schema, and it must match the signal emitter called `something` in the `foo` plugin. The 
+`$.steps.foo.signals.something` expression refers to the signal channel.
+
+### Closing signal channels
+
+Plugins can also close signal channels, which will notify all listening plugins that there are no more signals
+coming. If the plugin terminates, the signal channel is closed automatically.
+
+### Deployment and input information
+
+The deployment and input of a plugin constitutes a special signal listener that receives exactly one signal of a kind.
+If more than one signal is sent on the signal channel, the consecutive messages are ignored.
+
+However, a workflow author may, with the help of an expression, collect all signals in a channel into a list. This will
+delay the evaluation of that expression until the signal channel is closed and all messages of a kind are received.
+
+For example:
+
+```yaml
+steps:
+  foo:
+    ...
+  bar:
+    input:
+      baz: !expr collect($.steps.foo.signals.something)
+```
+
+### Outputs
+
+The current normal output results will also be transformed into signals, which depend on the completion of a plugin.
+However, a plugin is now allowed to not declare an output and work with signals instead.
```

Comment:
I can't think of a specific issue (I might need more coffee), but on the surface allowing the case of "no output" feels like it could cause problems.

Reply:
That's fair. I just feel like the `output` constitutes a sort of "punctuation" for a plugin's successful end. Maybe that's not functionally necessary.